sw/device/silicon_creator/lib/crypto/rsa_3072/rsa_3072.s

```diff
@@ -438,6 +438,372 @@ modexp_var_3072_f4:
+   *
+   * If aa < M, it follows that the carry bit aa[3072] = 0 (since M < 2^3072).
+   * It also follows that the borrow from subtracting M must be 1. In this
+   * case, select aa; otherwise, select aa-M.
+   */
+
+  /* x2 <= (!x2) & x3 */
+  xori      x2, x2, 1
+  and       x2, x2, x3
+
+  /* select aa if x2 = 0, otherwise aa-M */
+  bne       x2, x0, sel_aa
+
+  /* copy subtraction result to w4:w15 */
+  li        x8, 4
+  li        x11, 16
+  loopi     12, 2
+    bn.movr   x8, x11++
+    addi      x8, x8, 1
+
+  sel_aa:
+
+  ret
+
+/**
+ * Computes the R^2 Montgomery constant and stores it in DMEM.
+ *
+ *   Returns RR = (2^3072)^2 mod M
+ *
+ * This implementation is based on section 2.3 of "Montgomery Arithmetic from a
+ * Software Perspective" (https://eprint.iacr.org/2017/1057). For the purposes
+ * of RSA 3072, the parameters w and n from the paper are fixed (w=256, n=12).
+ *
+ * The algorithm from the paper assumes that 2^wn-1 <= M < 2^wn; we lightly
+ * adapt the implementation to accept any positive M by subtracting M from c0
+ * until c0 < M.
+ *
+ * The result is stored in dmem[in_rr]. This routine runs in variable time.
+ *
+ * Flags: Flags have no meaning beyond the scope of this subroutine.
+ *
+ * @param[in]  dmem[in_mod] pointer to first limb of modulus M in dmem
+ *
+ * clobbered registers: x2, x3, x7 to x11, x16, w2 to w27, w31
+ * clobbered Flag Groups: FG0
+ */
+ .globl precomp_rr
+precomp_rr:
+  /* prepare all-zero reg */
+  bn.xor    w31, w31, w31
+
+  /* set pointer to modulus */
+  la        x16, in_mod
+
+  /* zero [w4:w15] */
+  li        x9, 4
+  li        x10, 31
+  loopi     12, 1
+    bn.movr   x9++, x10
+
+  /* w16 <= 1 */
+  bn.addi   w16, w31, 1
+
+  /* Initialize c0
+     c0 = [w4:w15] <= [w4:w16] >> 1 = 2^3701 */
+  bn.rshi   w15, w16, w15 >> 1
+
+
+precomp_rr_sub_start:
+
+  /* Repeatedly subtract M until c0 < M
+       [w16:w27] <= (c0 - M) */
+  jal       x1, subtract_modulus_var
+
+  /* extract borrow bit from flags register */
+  csrrs     x2, 0x7c0, x0
+  andi      x2, x2, 1
+
+  /* if borrow is set, then the subtraction underflowed, meaning c0 < M; done */
+  bne       x2, x0, precomp_rr_sub_done
+
+  /* if we got here, then c0 > M; set c0 = c0 - M and repeat.
+       c0 = [w4:w15] <= [w16:w27] = c0 - M */
+  li        x8, 4
+  li        x11, 16
+  loopi     12, 2
+    bn.movr   x8, x11++
+    addi      x8, x8, 1
+
+  /* jump back to start of subtractions */
+  beq       x0, x0, precomp_rr_sub_start
+
+precomp_rr_sub_done:
+
+  /* Now, we know that c0 = [w4:w15] = (2^3071) mod M */
+
+  /* One modular doubling to get c1 \equiv 2^3072 mod M
+     c1 = [w4:w15] <= ([w4:w15] * 2) mod M = (2^3072) mod M */
+  jal     x1, double_mod_var
+
+  /* Compute (2^3072)^2 mod M by performing 3072 modular doublings.
+     Loop is nested only because #iterations must be < 1024 */
+  loopi     12, 4
+    loopi     256, 2
+      jal     x1, double_mod_var
+      nop
```

Comment:
Do you think it's worth noting that these `nop`s are necessary because `loopi` needs to take an even immediate or something like that? (Been a while since I looked at the OTBN ISA closely.)

Reply:
It's because `loopi` can't end in a jump or another loop; I'll add a comment.

Reply:
Throwing peanuts from the sidelines... Another option would be to use `loop` instead, which would allow you to construct the `12 * 256` loop count without eating a level of the loop stack.

Reply:
That's probably a good idea, thanks! If I need to update this again I'll implement it, otherwise will do it in a follow-up.